File: Stage_3/SIDNy/Lorenz_SINDy.py
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter
from sklearn.preprocessing import StandardScaler
import pysindy as ps

script_dir = Path(__file__).resolve().parent
data_path = script_dir / "Lorenz_Official" / "csv" / "train" / "X1train.csv"
time_path = script_dir / "Lorenz_Official" / "csv" / "train" / "X1train_timesteps.csv"

if not data_path.is_file():
    raise FileNotFoundError(f"Could not find dataset at: {data_path}")
if not time_path.is_file():
    raise FileNotFoundError(f"Could not find dataset at: {time_path}")

data = pd.read_csv(data_path, header=None)
t = pd.read_csv(time_path, header=None)

x = data.iloc[:, 0].to_numpy()
y = data.iloc[:, 1].to_numpy()
z = data.iloc[:, 2].to_numpy()
t = t.iloc[:, 0].to_numpy().flatten()

# Center and scale each variable to reduce numerical issues.
x = (x - x.mean()) / x.std()
y = (y - y.mean()) / y.std()
z = (z - z.mean()) / z.std()

X = np.column_stack((x, y, z))

# Smooth data and estimate derivatives.
window = 11
poly = 3
dt = float(np.mean(np.diff(t)))

x_s = savgol_filter(x, window_length=window, polyorder=poly)
y_s = savgol_filter(y, window_length=window, polyorder=poly)
z_s = savgol_filter(z, window_length=window, polyorder=poly)

Xs = np.column_stack((x_s, y_s, z_s))

Xdot = np.column_stack((
    savgol_filter(x_s, window_length=window, polyorder=poly, deriv=1, delta=dt),
    savgol_filter(y_s, window_length=window, polyorder=poly, deriv=1, delta=dt),
    savgol_filter(z_s, window_length=window, polyorder=poly, deriv=1, delta=dt),
))

# Normalize inputs/targets for the optimizer.
scaler_X = StandardScaler()
scaler_dX = StandardScaler()
Xn = scaler_X.fit_transform(Xs)
Xdot_n = scaler_dX.fit_transform(Xdot)

print("X shape:", Xn.shape)
print("Xdot shape:", Xdot_n.shape)
print("X sample:\n", Xn[:5])
print("Xdot sample:\n", Xdot_n[:5])

# Build a stable SINDy model.
library = ps.PolynomialLibrary(degree=2, include_bias=True)
optimizer = ps.STLSQ(threshold=0.05, max_iter=100)
model = ps.SINDy(
    feature_library=library,
    optimizer=optimizer,
)
model.fit(Xn, t=t, x_dot=Xdot_n)
model.print()

# Simulate with a robust solver.
X0 = Xn[0]
x_model_n = np.asarray(model.simulate(
    X0,
    t,
    integrator_kws={
        "method": "RK45",
        "rtol": 1e-8,
        "atol": 1e-10,
    },
))

# Convert back to physical coordinates.
x_model = scaler_X.inverse_transform(x_model_n)

fig, axs = plt.subplots(3, 1, figsize=(10, 12), sharex=True)
axs[0].plot(t, X[:, 0], 'r', label='True x')
axs[0].plot(t, x_model[:, 0], 'b--', label='SINDy x')
axs[0].set_ylabel('x')
axs[0].legend()
axs[0].grid(True)

axs[1].plot(t, X[:, 1], 'r', label='True y')
axs[1].plot(t, x_model[:, 1], 'b--', label='SINDy y')
axs[1].set_ylabel('y')
axs[1].legend()
axs[1].grid(True)

axs[2].plot(t, X[:, 2], 'r', label='True z')
axs[2].plot(t, x_model[:, 2], 'b--', label='SINDy z')
axs[2].set_xlabel('Time (s)')
axs[2].set_ylabel('z')
axs[2].legend()
axs[2].grid(True)

plt.tight_layout()
plt.show(block=True)

# Fit percentage (using the same scaled data used for fitting).
fit_percentage = (
    1 - np.linalg.norm(X - x_model, ord='fro') / np.linalg.norm(X - np.mean(X, axis=0))
)
print(f"Fit percentage: {fit_percentage * 100:.2f}%")

File: Stage_3/SIDNy/Lorenz_SINDy.m
clc;
clear;
close all;

%% Load data
baseDir = fileparts(mfilename('fullpath'));
dataFile = fullfile(baseDir, 'Lorenz_Official', 'csv', 'train', 'X1train.csv');
timeFile = fullfile(baseDir, 'Lorenz_Official', 'csv', 'train', 'X1train_timesteps.csv');

data = readmatrix(dataFile);
timesteps = readmatrix(timeFile);

X = data;
t = timesteps(:,1);

x = X(:,1);
y = X(:,2);
z = X(:,3);

%% Remove mean and normalize
% x = x - mean(x);
% y = y - mean(y);
% z = z - mean(z);
% 
% x = x / std(x);
% y = y / std(y);
% z = z / std(z);

%% Plot the data
figure

subplot(3,1,1)
plot(t,x)
grid on
xlabel('Time (s)')
ylabel('x1')
title('x1 vs Time')

subplot(3,1,2)
plot(t,y)
grid on
xlabel('Time (s)')
ylabel('x2')
title('x2 vs Time')

subplot(3,1,3)
plot(t,z)
grid on
xlabel('Time (s)')
ylabel('x3')
title('x3 vs Time')

%% prepare the data and derivation matrices
dt = mean(diff(t));

window = 11;
poly = 3;

x_s = sgolayfilt(x, poly, window);
y_s = sgolayfilt(y, poly, window);
z_s = sgolayfilt(z, poly, window);

dt = mean(diff(t));

dx = sgolayfilt(x_s, poly, window, [], 1) / dt;
dy = sgolayfilt(y_s, poly, window, [], 1) / dt;
dz = sgolayfilt(z_s, poly, window, [], 1) / dt;

X = [x_s, y_s, z_s];

mu = mean(X);
sigma = std(X);

Xn = (X - mu) ./ sigma;

Xdot = [dx, dy, dz];
Xdot_n = (Xdot - mean(Xdot)) ./ std(Xdot);


%% Build Library matrix

Theta = [
    ones(size(Xn,1),1), ...
    Xn(:,1), Xn(:,2), Xn(:,3), ...
    Xn(:,1).*Xn(:,2), ...
    Xn(:,1).*Xn(:,3), ...
    Xn(:,2).*Xn(:,3)
];

%% Solve for Sparse
lambda = 0.05;
Xi = Theta \ Xdot;

for k = 1:10
    small = abs(Xi) < lambda;
    Xi(small) = 0;

    for i = 1:size(Xdot,2)
        idx = ~small(:,i);
        Xi(idx,i) = Theta(:,idx) \ Xdot(:,i);
    end
end

%% Obtain Equations
terms = {'1','x','y','z','xy','xz','yz'};

for state = 1:3
    fprintf('\n xdot_%d = ', state)

    for k = 1:length(terms)
        if abs(Xi(k,state)) > 1e-8
            fprintf(' %+f*%s ', Xi(k,state), terms{k});
        end
    end

    fprintf('\n')
end

%% Results Simulations
x_sim = zeros(length(t),3);
x_sim(1,:) = X(1,:);

for k = 1:length(t)-1

    s = x_sim(k,:)';   % 3×1 column vector

    f = @(v) [
        Xi(1,1) + Xi(2,1)*v(1) + Xi(3,1)*v(2) + Xi(4,1)*v(3) + Xi(5,1)*v(1)*v(2) + Xi(6,1)*v(1)*v(3) + Xi(7,1)*v(2)*v(3);
        Xi(1,2) + Xi(2,2)*v(1) + Xi(3,2)*v(2) + Xi(4,2)*v(3) + Xi(5,2)*v(1)*v(2) + Xi(6,2)*v(1)*v(3) + Xi(7,2)*v(2)*v(3);
        Xi(1,3) + Xi(2,3)*v(1) + Xi(3,3)*v(2) + Xi(4,3)*v(3) + Xi(5,3)*v(1)*v(2) + Xi(6,3)*v(1)*v(3) + Xi(7,3)*v(2)*v(3)
    ];

    k1 = f(s);
    k2 = f(s + 0.5*dt*k1);
    k3 = f(s + 0.5*dt*k2);
    k4 = f(s + dt*k3);

    increment = (dt/6)*(k1 + 2*k2 + 2*k3 + k4);   % 3×1

    x_sim(k+1,:) = (s + increment)';  % convert back to 1×3
end

% x Comparison
figure
plot(t,X(:,1),'LineWidth',1.5)
hold on
plot(t,x_sim(:,1),'--','LineWidth',1.5)
grid on
legend('True','SINDy')
xlabel('Time')
ylabel('x')
title('x Comparison')

% y Comparison
figure
plot(t,X(:,2),'LineWidth',1.5)
hold on
plot(t,x_sim(:,2),'--','LineWidth',1.5)
grid on
legend('True','SINDy')
xlabel('Time')
ylabel('y')
title('y Comparison')

% Z Comparison
figure
plot(t,X(:,3),'LineWidth',1.5)
hold on
plot(t,x_sim(:,3),'--','LineWidth',1.5)
grid on
legend('True','SINDy')
xlabel('Time')
ylabel('z')
title('z Comparison')

%% Phase Portrait Comparison
figure
plot3(X(:,1),X(:,2),X(:,3))
hold on
plot3(x_sim(:,1),x_sim(:,2),x_sim(:,3),'--')
grid on
legend('True','SINDy')
xlabel('x'); ylabel('y'); zlabel('z')
title('Lorenz Attractor')

%% Quantative Fit
fit_percent = 100 * (1 - norm(X - x_sim,'fro') / norm(X,'fro'))